onnxscript/rewriter/onnxruntime/bfloat16_utils/bfloat16_converter.py

```diff
@@ -0,0 +1,84 @@
+import logging
+
+from onnxscript import ir
+
+logger = logging.getLogger(__name__)
+
+
+def _convert_inputs_from_bfloat16_to_float16(value: ir.Input):
+    if value.dtype == ir.DataType.BFLOAT16:
+        value.dtype = ir.DataType.FLOAT16
+        value_users = list(value.uses())
+        for node, index in value_users:
+            _insert_cast_nodes_for_float16_to_bfloat16_to_inputs(node, index)
+
+
+def _convert_outputs_from_bfloat16_to_float16(value: ir.Value):
+    if value.dtype == ir.DataType.BFLOAT16:
+        node = value.producer()
+        index = value.index()
+        if node is None or index is None:
+            logger.warning("Output value %s has no producer or index", value)
+            return
+        _insert_cast_nodes_for_bfloat16_to_float16_to_outputs(node, index)
+
+
+def _insert_cast_nodes_for_float16_to_bfloat16_to_inputs(node: ir.Node, index: int):
+    attr = ir.AttrInt64(name="to", value=ir.DataType.BFLOAT16)
+    cast = ir.Node(
+        domain=node.domain,
+        op_type="Cast",
+        inputs=[node.inputs[index]],
+        num_outputs=1,
+        attributes=[attr],
+    )
+    cast.outputs[0].dtype = ir.DataType.BFLOAT16  # type: ignore[union-attr]
+    cast.outputs[0].shape = node.inputs[index].shape  # type: ignore[union-attr]
+    node.prepend(cast)
+    node.replace_input_with(index, cast.outputs[0])
+
+
+def _insert_cast_nodes_for_bfloat16_to_float16_to_outputs(node: ir.Node, index: int):
+    attr = ir.AttrInt64(name="to", value=ir.DataType.FLOAT16)
+    cast = ir.Node(
+        domain=node.domain,
+        op_type="Cast",
+        inputs=[node.outputs[index]],
+        num_outputs=1,
+        attributes=[attr],
+    )
+    cast.outputs[0].dtype = ir.DataType.FLOAT16
+    cast.outputs[0].shape = node.outputs[index].shape
```

Comment:
We may need to retain the output names. Potentially rename the original value so that it doesn't collide?

Reply:
Swap the output value names you meant?

Reply:
> Swap the output value names you meant?

That works

Reply:
What difference does it make? The newly created node is not the same naming authority as the graph? If that's the case, Isn't swapping making no difference in terms of the duplication of names in the same graph?

Reply:
So I was thinking the output names need to be preserved. For example, consider

```
v_n -> output 
```

where output is graph.output[0]. Then after adding cast we have

```
v_n -> output -> v0
```

where v0 is the name of the new graph output

we should now rename "v0" to "output", so that the graph output name stays unchanged. Like you suggested swapping the names may be sufficient:

```
v_n -> v0 -> output
```

otherwise we can also rename output->output_bfloat16 and v0->output